src/operator/src/statement/copy_table_from.rs

```diff
@@ -82,49 +117,62 @@ impl StatementExecutor {
         Ok((object_store, entries))
     }
 
-    async fn infer_schema(
+    async fn collect_metadata(
         &self,
-        format: &Format,
-        object_store: ObjectStore,
-        path: &str,
-    ) -> Result<SchemaRef> {
+        object_store: &ObjectStore,
+        format: Format,
+        path: String,
+    ) -> Result<FileMetadata> {
         match format {
-            Format::Csv(format) => Ok(Arc::new(
-                format
-                    .infer_schema(&object_store, path)
-                    .await
-                    .context(error::InferSchemaSnafu { path })?,
-            )),
-            Format::Json(format) => Ok(Arc::new(
-                format
-                    .infer_schema(&object_store, path)
-                    .await
-                    .context(error::InferSchemaSnafu { path })?,
-            )),
+            Format::Csv(format) => Ok(FileMetadata::Csv {
+                schema: Arc::new(
+                    format
+                        .infer_schema(object_store, &path)
+                        .await
+                        .context(error::InferSchemaSnafu { path: &path })?,
+                ),
+                format,
+                path,
+            }),
+            Format::Json(format) => Ok(FileMetadata::Json {
+                schema: Arc::new(
+                    format
+                        .infer_schema(object_store, &path)
+                        .await
+                        .context(error::InferSchemaSnafu { path: &path })?,
+                ),
+                format,
+                path,
+            }),
             Format::Parquet(_) => {
-                let reader = object_store
-                    .reader(path)
+                let mut reader = object_store
+                    .reader(&path)
```

Comment:
Is there no need for Read Buffer here?

Reply:
Yes, the reader buffer is not helpful if it reads metadata starting from the tailing.